proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1223 @@
+  ///
+  /// `unsafeBytes` must be correctly aligned for accessing
+  /// an element of type `Element`, and must contain a number of bytes
+  /// that is an exact multiple of `Element`'s stride.
+  ///
+  /// - Parameters:
+  ///   - unsafeBytes: a buffer to initialized elements.
+  ///   - type: the type to use when interpreting the bytes in memory.
+  ///   - owner: a binding whose lifetime must exceed that of
+  ///            the newly created `Span`.
+  public init<Owner: ~Copyable & ~Escapable>(
+    unsafeBytes buffer: UnsafeRawBufferPointer,
+    owner: borrowing Owner
+  )
+
+  /// Unsafely create a `Span` over a span of initialized memory.
+  ///
+  /// The memory representing `count` instances starting at
+  /// `unsafeRawPointer` must be owned by the instance `owner`,
+  /// meaning that as long as `owner` is alive the memory will remain valid.
+  ///
+  /// `unsafeRawPointer` must be correctly aligned for accessing
+  /// an element of type `Element`.
+  ///
+  /// - Parameters:
+  ///   - unsafeRawPointer: a pointer to the first initialized element.
+  ///   - type: the type to use when interpreting the bytes in memory.
+  ///   - count: the number of initialized elements in the view.
+  ///   - owner: a binding whose lifetime must exceed that of
+  ///            the newly created `Span`.
+  public init<Owner: ~Copyable & ~Escapable>(
+    unsafeStart pointer: UnsafeRawPointer,
+    byteCount: Int,
+    owner: borrowing Owner
+  )
+}
+```
+
+##### Basic API:
+
+The following properties, functions and subscripts have direct counterparts in the `Collection` protocol hierarchy. Their semantics shall be as described where they counterpart is declared (in `Collection` or `RandomAccessCollection`). The only difference with their counterpart should be a lifetime dependency annotation where applicable, allowing them to return borrowed nonescapable values or borrowed noncopyable values.
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public var count: Int { get }
+  public var isEmpty: Bool { get }
+
+  public subscript(_ position: Int) -> Element { _read }
+}
+```
+
+##### Accessing subranges of elements:
+
+In SE-0437, `UnsafeBufferPointer`'s slicing subscript was replaced by the `extracting(_ bounds:)` functions, due to the copyability assumption baked into the standard library's `Slice` type. `Span` has similar requirements as `UnsafeBufferPointer`, in that it must be possible to obtain an instance representing a subrange of the same memory as another `Span`. We therefore follow in the footsteps of SE-0437, and add a family of `extracting()` functions:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  /// Constructs a new span over the items within the supplied range of
+  /// positions within this span.
+  ///
+  /// The returned span's first item is always at offset 0; unlike buffer
+  /// slices, extracted spans do not generally share their indices with the
```

Comment:
What does the word `generally` mean here? From a caller's perspective, I assume they should just fully expect that the index is not shared at all, right?

Reply:
Yeah that's not very good. I adapted this from the new `extracting` functions on `UnsafeBufferPointer`, which use similar words. If you extract a sub-span with the first N items of a span, then the first N indices are effectively shareable, but that's an obvious coincidence, and we don't need to accommodate it by saying "generally".